modules/snooker/src/main/ts/ephox/snooker/api/Warehouse.ts

```diff
@@ -80,42 +80,47 @@ const generate = <T extends Structs.Detail>(list: Structs.RowDetail<T>[]): Wareh
-      columns = generateColumns<T>(rowData);
-      colgroups = [ Structs.colgroup(rowData.element as SugarElement<HTMLTableColElement>, Obj.values(columns)) ];
-    } else {
-      const currentRow: Structs.DetailExt[] = [];
-      Arr.each(rowData.cells, (rowCell) => {
-        let start = 0;
-
-        // If this spot has been taken by a previous rowspan, skip it.
-        while (access[key(rowCount, start)] !== undefined) {
-          start++;
+  const { pass: colgroupRows, fail: rows } = Arr.partition(list, (rowData) => rowData.section === 'colgroup');
+
+  // Handle rows first
+  Arr.each(rows, (rowData) => {
+    const currentRow: Structs.DetailExt[] = [];
+    Arr.each(rowData.cells, (rowCell) => {
+      let start = 0;
+
+      // If this spot has been taken by a previous rowspan, skip it.
+      while (access[key(rowCount, start)] !== undefined) {
+        start++;
+      }
+
+      const isLocked = Obj.hasNonNullableKey(lockedColumns, start.toString());
+      const current = Structs.extended(rowCell.element, rowCell.rowspan, rowCell.colspan, rowCount, start, isLocked);
+
+      // Occupy all the (row, column) positions that this cell spans for.
+      for (let occupiedColumnPosition = 0; occupiedColumnPosition < rowCell.colspan; occupiedColumnPosition++) {
+        for (let occupiedRowPosition = 0; occupiedRowPosition < rowCell.rowspan; occupiedRowPosition++) {
+          const rowPosition = rowCount + occupiedRowPosition;
+          const columnPosition = start + occupiedColumnPosition;
+          const newpos = key(rowPosition, columnPosition);
+          access[newpos] = current;
+          maxColumns = Math.max(maxColumns, columnPosition + 1);
         }
+      }
 
-        const isLocked = Obj.hasNonNullableKey(lockedColumns, start.toString());
-        const current = Structs.extended(rowCell.element, rowCell.rowspan, rowCell.colspan, rowCount, start, isLocked);
-
-        // Occupy all the (row, column) positions that this cell spans for.
-        for (let occupiedColumnPosition = 0; occupiedColumnPosition < rowCell.colspan; occupiedColumnPosition++) {
-          for (let occupiedRowPosition = 0; occupiedRowPosition < rowCell.rowspan; occupiedRowPosition++) {
-            const rowPosition = rowCount + occupiedRowPosition;
-            const columnPosition = start + occupiedColumnPosition;
-            const newpos = key(rowPosition, columnPosition);
-            access[newpos] = current;
-            maxColumns = Math.max(maxColumns, columnPosition + 1);
-          }
-        }
+      currentRow.push(current);
+    });
 
-        currentRow.push(current);
-      });
+    maxRows++;
+    cells.push(Structs.rowdetail(rowData.element, currentRow, rowData.section));
+    rowCount++;
+  });
 
-      maxRows++;
-      cells.push(Structs.rowdetail(rowData.element, currentRow, rowData.section));
-      rowCount++;
-    }
+  // Handle colgroups
+  // Note: Currently only a single colgroup is supported so just use the last one
```

Comment:
Your change here breaks this specific implementation, so please make sure to revert this part of the change so we are guaranteed to only use the last `colgroup` element.

- [ ] Revert doing colgroups.push

Reply:
Have reworked it based on DMs in Slack